wuffs: protect against crafted images that cause overflows

jcollie opened the pull request:
Fixes #9579

Protect against panics caused by integer overflows by using functions that allow integer overflows to be caught instead of causing a panic.

Also protect against DOS from images that might not cause an overflow but do consume an absurd amount of memory by limiting images to a maximum size of 4GiB.

mitchellh reviewed (changes requested):
One change requested. 

Is 8192 arbitrary or does Kitty also reject images of a certain size? Might be useful to see if there is some limit there we can match.

jcollie committed:
wuffs: protect against crafted images that cause overflows

Fixes #9579

Protect against panics caused by integer overflows by using functions
that allow integer overflows to be caught instead of causing a panic.

Also protect against DOS from images that might not cause an
overflow but do consume an absurd amount of memory by limiting
images to a maximum size of 4GiB (which is the maximum size of
`image-storage-limit`).

jcollie commented on pkg/wuffs/src/main.zig:
```diff
@@ -5,6 +5,9 @@ pub const jpeg = @import("jpeg.zig");
 pub const Error = @import("error.zig").Error;
 
+pub const maximum_image_width = 8192;
+pub const maximum_image_height = 8192;
```
Switch to limiting images to a maximum decoded size of 4GiB.

mitchellh commented:
I want to note it might also be worth changing this part:

```
width * height * @sizeOf(c.wuffs_base__color_u32_argb_premul),
```

To use `std.math.mul` so that we can catch overflows and just raise them appropriately. So even if we change the max image size in the future, we can still prevent integer overflows.
